src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,343 @@
+}
+
+//! Evaluate cost over whole dataset.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates)
+{
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  // Re-calculate impostors on transformed dataset.
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate cost due to distance between target neighbors & data point.
+      double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i)));
+      cost += (1 - regularization) * eval;
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+         double eval = metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(targetNeighbors(j, i))) -
+                       metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+};
+
+//! Calculate cost over batches.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates,
+                                          const size_t begin,
+                                          const size_t batchSize)
+{
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
+
+  for (size_t i = begin; i < begin + batchSize; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate cost due to distance between target neighbors & data point.
+      double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i)));
+      cost += (1 - regularization) * eval;
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+         double eval = metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(targetNeighbors(j, i))) -
+                       metric.Evaluate(transformedDataset.col(i),
+                            transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+}
+
+//! Compute gradient over whole dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& coordinates,
+                                        arma::mat& gradient)
```

Comment:
This assumes the same impostors as `Evaluate()` (or, I guess, the real assumption is that `Gradient()` will always be called after `Evaluate()` with the same `coordinates`).  I think that is true in most cases, but if you are having trouble getting the algorithm to converge correctly, you could always try manually recalculating impostors here.  In addition, if you implement an `EvaluateWithGradient()` function, an optimizer can make use of that and you can save on the calculation of impostors there.  (See http://mlpack.org/docs/mlpack-git/doxygen/optimizertutorial.html, section "The FunctionType API", if you haven't already)